File: runtime/schema/typescript/src/runtime/parse/walk.ts
/**
 * Per-field parse walker. Walks each field of a TypeDef, applies defaults when
 * absent, coerces primitives, runs normalize and parse hooks, and recurses into
 * nested types / arrays. Strict mode rejects unknown fields with
 * {validator: "unknown_field"} keyed at the field path.
 */

import {
  addFieldError,
  addNestedErrors,
  hasErrors,
  newValidationErrors,
  setFieldErrors,
  type ValidationErrors,
} from 'superscalar/validation';
import { isAnyJSONScalar, jsonShapeOf, structuredJSONType } from '../validation/types';
import type { FieldDef, ScalarDef, Schema, TypeDef, TypeRef } from '../validation/types';
import { coerceBool, coerceFloat, coerceInt } from './coerce';
import { applyDefault } from './defaults';
import type { ScalarNormalizeRegistry, ScalarParseRegistry } from './registry';

const BUILTIN_SCALARS = new Set(['String', 'Int', 'Float', 'Boolean', 'ID']);

type RefKind = 'scalar' | 'enum' | 'type' | 'input' | 'builtin';

export interface WalkContext {
  schema: Schema;
  parseRegistry: ScalarParseRegistry;
  normalizeRegistry: ScalarNormalizeRegistry;
  strict: boolean;
}

export function walkTypeDef(
  ctx: WalkContext,
  td: TypeDef,
  raw: Record<string, unknown>
): { data: Record<string, unknown>; errors: ValidationErrors } {
  const result: Record<string, unknown> = {};
  const errors = newValidationErrors();
  const known = new Set<string>();
  for (const field of td.fields) {
    const key = fieldKey(field);
    known.add(key);
    walkField(ctx, field, key, raw, result, errors);
  }
  if (ctx.strict) {
    for (const k of Object.keys(raw)) {
      if (known.has(k)) continue;
      addFieldError(errors, k, 'unknown_field', 'unknown field');
    }
  }
  return { data: result, errors };
}

function walkField(
  ctx: WalkContext,
  field: FieldDef,
  key: string,
  raw: Record<string, unknown>,
  result: Record<string, unknown>,
  errors: ValidationErrors
): void {
  const present = Object.prototype.hasOwnProperty.call(raw, key);
  const value = present ? raw[key] : undefined;

  if (!present) {
    if (field.defaultValue === null || field.defaultValue === undefined) return;
    if (field.typeRef.isArray) return;
    const kind = resolveRefKind(ctx.schema, field.typeRef);
    if (kind === 'type' || kind === 'input') return;
    const scalar = kind === 'scalar' ? scalarDefFor(ctx.schema, field.typeRef.name) : null;
    const [v, ok] = applyDefault(field, scalar);
    if (!ok) {
      addFieldError(errors, key, 'default', 'default value is not parseable for declared type');
      return;
    }
    result[key] = v;
    return;
  }

  if (value === null) {
    result[key] = null;
    return;
  }

  const kind = resolveRefKind(ctx.schema, field.typeRef);
  if (field.typeRef.isArray) {
    walkArrayField(ctx, field, key, kind, value, result, errors);
    return;
  }
  walkSingleField(ctx, field, key, kind, value, result, errors);
}

function walkSingleField(
  ctx: WalkContext,
  field: FieldDef,
  key: string,
  kind: RefKind,
  value: unknown,
  result: Record<string, unknown>,
  errors: ValidationErrors
): void {
  switch (kind) {
    case 'scalar': {
      const scalar = scalarDefFor(ctx.schema, field.typeRef.name);
      if (!scalar) {
        result[key] = value;
        return;
      }
      const [out, ok] = applyScalar(ctx, scalar, value);
      if (!ok) {
        addFieldError(errors, key, 'type', typeMismatchMessage(scalar));
        result[key] = value;
        return;
      }
      const finalValue = applyScalarParse(ctx, scalar, out, key, errors);
      result[key] = finalValue;
      return;
    }
    case 'enum':
      result[key] = value;
      return;
    case 'builtin': {
      const [out, ok] = applyBuiltin(field.typeRef.name, value, ctx.strict);
      if (!ok) {
        addFieldError(errors, key, 'type', `expected ${field.typeRef.name} value`);
        result[key] = value;
        return;
      }
      result[key] = out;
      return;
    }
    case 'type': {
      const td = ctx.schema.types ? ctx.schema.types[field.typeRef.name] : undefined;
      if (!td || !isObject(value)) {
        addFieldError(errors, key, 'type', 'expected object value');
        result[key] = value;
        return;
      }
      const nested = walkTypeDef(ctx, td, value);
      if (hasErrors(nested.errors)) {
        addNestedErrors(errors, key, nested.errors);
      }
      result[key] = nested.data;
      return;
    }
    case 'input': {
      const td = ctx.schema.inputs ? ctx.schema.inputs[field.typeRef.name] : undefined;
      if (!td || !isObject(value)) {
        addFieldError(errors, key, 'type', 'expected object value');
        result[key] = value;
        return;
      }
      const nested = walkTypeDef(ctx, td, value);
      if (hasErrors(nested.errors)) {
        addNestedErrors(errors, key, nested.errors);
      }
      result[key] = nested.data;
      return;
    }
  }
}

function walkArrayField(
  ctx: WalkContext,
  field: FieldDef,
  key: string,
  kind: RefKind,
  value: unknown,
  result: Record<string, unknown>,
  errors: ValidationErrors
): void {
  if (!Array.isArray(value)) {
    addFieldError(errors, key, 'type', 'expected array value');
    result[key] = value;
    return;
  }
  if (!field.typeRef.isArrayOfArrays) {
    result[key] = walkArrayElements(ctx, field, key, kind, value, errors);
    return;
  }
  // T[][]: walk each inner list; its elements report at key[i][j]. A null
  // inner list passes through for validation to reject, as a null element
  // of T[] does.
  const out: unknown[] = new Array(value.length);
  for (let i = 0; i < value.length; i += 1) {
    const inner = value[i];
    const innerKey = `${key}[${i}]`;
    if (inner === null || inner === undefined) {
      out[i] = null;
      continue;
    }
    if (!Array.isArray(inner)) {
      addFieldError(errors, innerKey, 'type', 'expected an array');
      out[i] = inner;
      continue;
    }
    out[i] = walkArrayElements(ctx, field, innerKey, kind, inner, errors);
  }
  result[key] = out;
}

function walkArrayElements(
  ctx: WalkContext,
  field: FieldDef,
  key: string,
  kind: RefKind,
  value: unknown[],
  errors: ValidationErrors
): unknown[] {
  const out: unknown[] = new Array(value.length);
  for (let i = 0; i < value.length; i += 1) {
    const elem = value[i];
    const elemKey = `${key}[${i}]`;
    if (elem === null || elem === undefined) {
      out[i] = null;
      continue;
    }
    switch (kind) {
      case 'scalar': {
        const scalar = scalarDefFor(ctx.schema, field.typeRef.name);
        if (!scalar) {
          out[i] = elem;
          continue;
        }
        const [v, ok] = applyScalar(ctx, scalar, elem);
        if (!ok) {
          addFieldError(errors, elemKey, 'type', typeMismatchMessage(scalar));
          out[i] = elem;
          continue;
        }
        out[i] = applyScalarParse(ctx, scalar, v, elemKey, errors);
        continue;
      }
      case 'enum':
        out[i] = elem;
        continue;
      case 'builtin': {
        const [v, ok] = applyBuiltin(field.typeRef.name, elem, ctx.strict);
        if (!ok) {
          addFieldError(errors, elemKey, 'type', `expected ${field.typeRef.name} value`);
          out[i] = elem;
          continue;
        }
        out[i] = v;
        continue;
      }
      case 'type': {
        const td = ctx.schema.types ? ctx.schema.types[field.typeRef.name] : undefined;
        if (!td || !isObject(elem)) {
          addFieldError(errors, elemKey, 'type', 'expected object value');
          out[i] = elem;
          continue;
        }
        const nested = walkTypeDef(ctx, td, elem);
        if (hasErrors(nested.errors)) {
          addNestedErrors(errors, elemKey, nested.errors);
        }
        out[i] = nested.data;
        continue;
      }
      case 'input': {
        const td = ctx.schema.inputs ? ctx.schema.inputs[field.typeRef.name] : undefined;
        if (!td || !isObject(elem)) {
          addFieldError(errors, elemKey, 'type', 'expected object value');
          out[i] = elem;
          continue;
        }
        const nested = walkTypeDef(ctx, td, elem);
        if (hasErrors(nested.errors)) {
          addNestedErrors(errors, elemKey, nested.errors);
        }
        out[i] = nested.data;
        continue;
      }
    }
  }
  return out;
}

function applyScalar(
  ctx: WalkContext,
  scalar: ScalarDef,
  value: unknown
): [unknown, boolean] {
  // Any JSON value is one, whatever the scalar's primitive; validation
  // checks it.
  if (isAnyJSONScalar(scalar)) return [value, true];
  // A JSON object or array scalar takes its object or array, or the value's
  // JSON text, which the String primitive's steps below read.
  const shape = structuredJSONType(scalar);
  if (shape) {
    if (jsonShapeOf(value) === shape) return [value, true];
    if (typeof value !== 'string') return [value, false];
  }
  switch (scalar.primitive) {
    case 'Int':
      return coerceInt(value, ctx.strict);
    case 'Float':
      return coerceFloat(value, ctx.strict);
    case 'Boolean':
      return coerceBool(value, ctx.strict);
    case 'String':
    case '': {
      if (typeof value !== 'string') return [value, false];
      let s = value;
      if (scalar.hasCustomNormalize) {
        const fn = ctx.normalizeRegistry.get(scalar.name);
        if (fn) s = fn(s);
      }
      return [s, true];
    }
    default:
      return [value, true];
  }
}

function applyScalarParse(
  ctx: WalkContext,
  scalar: ScalarDef,
  value: unknown,
  key: string,
  errors: ValidationErrors
): unknown {
  const shape = structuredJSONType(scalar);
  if (shape) return parseStructuredJSON(ctx, scalar, shape, value, key, errors);
  if (!scalar.hasCustomParse) return value;
  const fn = ctx.parseRegistry.get(scalar.name);
  if (!fn) return value;
  const input = typeof value === 'string' ? value : String(value);
  const [parsed, errs] = fn(input);
  if (errs.length > 0) {
    setFieldErrors(errors, key, errs);
    return value;
  }
  if (scalar.primitive === 'Int') {
    const [coerced, ok] = coerceInt(parsed, false);
    return ok ? coerced : value;
  }
  return parsed;
}

/**
 * Parses a value of a JSON object or array scalar (structuredJSONType) into
 * that object or array, the value every generated type holds. A string is
 * the value's JSON text. When the scalar has a registered parser
 * (Generic_StringMap), the text, or the JSON text of an object or array,
 * goes through it and its canonical text is decoded; a parser error is the
 * result. Text that does not decode to the declared shape is kept as it is,
 * for validation to report.
 */
function parseStructuredJSON(
  ctx: WalkContext,
  scalar: ScalarDef,
  shape: 'object' | 'array',
  value: unknown,
  key: string,
  errors: ValidationErrors
): unknown {
  let text = typeof value === 'string' ? value : undefined;
  const fn = scalar.hasCustomParse ? ctx.parseRegistry.get(scalar.name) : undefined;
  if (fn) {
    const input = text ?? JSON.stringify(value);
    if (input === undefined) return value;
    const [parsed, errs] = fn(input);
    if (errs.length > 0) {
      setFieldErrors(errors, key, errs);
      return value;
    }
    text = parsed;
  }
  if (text === undefined) return value;
  try {
    const decoded: unknown = JSON.parse(text);
    return jsonShapeOf(decoded) === shape ? decoded : value;
  } catch {
    return value;
  }
}

function applyBuiltin(name: string, value: unknown, strict: boolean): [unknown, boolean] {
  switch (name) {
    case 'Int':
      return coerceInt(value, strict);
    case 'Float':
      return coerceFloat(value, strict);
    case 'Boolean':
      return coerceBool(value, strict);
    case 'String':
    case 'ID':
      return typeof value === 'string' ? [value, true] : [value, false];
    default:
      return [value, true];
  }
}

function typeMismatchMessage(scalar: ScalarDef): string {
  const shape = structuredJSONType(scalar);
  if (shape) return `expected a JSON ${shape} or its JSON text`;
  switch (scalar.primitive) {
    case 'Int':
      return 'expected integer value';
    case 'Float':
      return 'expected numeric value';
    case 'Boolean':
      return 'expected boolean value';
    case 'String':
    case '':
      return 'expected string value';
    default:
      return 'type mismatch';
  }
}

function resolveRefKind(schema: Schema, ref: TypeRef): RefKind {
  if (schema.scalars && schema.scalars[ref.name]) return 'scalar';
  if (schema.enums && schema.enums[ref.name]) return 'enum';
  if (schema.types && schema.types[ref.name]) return 'type';
  if (schema.inputs && schema.inputs[ref.name]) return 'input';
  if (BUILTIN_SCALARS.has(ref.name)) return 'builtin';
  return 'builtin';
}

function scalarDefFor(schema: Schema, name: string): ScalarDef | null {
  if (!schema.scalars) return null;
  return schema.scalars[name] || null;
}

export function fieldKey(field: FieldDef): string {
  return field.jsonKey || field.name;
}

function isObject(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

File: runtime/schema/python/superschematic_schema_runtime/parse/walk.py
"""Per-field parse walker.

Applies defaults when absent, coerces primitives, runs normalize and parse
hooks, and recurses into nested types / arrays. Strict mode rejects unknown
fields with ``{"validator": "unknown_field"}``.
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any

from .. import ValidationError
from ..errors import (
    ValidationErrors,
    append_field_error,
    has_errors as _has_errors,
    new_validation_errors,
)
from ..validation.types import FieldDef, ScalarDef, Schema, TypeDef, TypeRef, json_shape_of
from .coerce import coerce_bool, coerce_float, coerce_int
from .defaults import apply_default
from .registry import ScalarNormalizeRegistry, ScalarParseRegistry


_BUILTIN_SCALARS = frozenset({"String", "Int", "Float", "Boolean", "ID"})


@dataclass
class WalkContext:
    schema: Schema
    parse_registry: ScalarParseRegistry
    normalize_registry: ScalarNormalizeRegistry
    strict: bool = False


def _is_object(value: Any) -> bool:
    return isinstance(value, dict)


def _resolve_ref_kind(schema: Schema, ref: TypeRef) -> str:
    if ref.name in schema.scalars:
        return "scalar"
    if ref.name in schema.enums:
        return "enum"
    if ref.name in schema.types:
        return "type"
    if ref.name in schema.inputs:
        return "input"
    return "builtin"


def field_key(field: FieldDef) -> str:
    return field.json_key or field.name


def _type_mismatch_message(scalar: ScalarDef) -> str:
    shape = scalar.structured_json_type()
    if shape:
        return f"expected a JSON {shape} or its JSON text"
    primitive = scalar.primitive
    if primitive == "Int":
        return "expected integer value"
    if primitive == "Float":
        return "expected numeric value"
    if primitive == "Boolean":
        return "expected boolean value"
    if primitive in ("String", ""):
        return "expected string value"
    return "type mismatch"


def _apply_scalar(ctx: WalkContext, scalar: ScalarDef, value: Any) -> tuple[Any, bool]:
    # Any JSON value is one, whatever the scalar's primitive; validation
    # checks it.
    if scalar.is_any_json():
        return value, True
    # A JSON object or array scalar takes its dict or list, or the value's
    # JSON text, which the String primitive's steps below read.
    shape = scalar.structured_json_type()
    if shape:
        if json_shape_of(value) == shape:
            return value, True
        if not isinstance(value, str):
            return value, False
    primitive = scalar.primitive
    if primitive == "Int":
        return coerce_int(value, ctx.strict)
    if primitive == "Float":
        return coerce_float(value, ctx.strict)
    if primitive == "Boolean":
        return coerce_bool(value, ctx.strict)
    if primitive in ("String", ""):
        if not isinstance(value, str):
            return value, False
        s = value
        if scalar.has_custom_normalize:
            fn = ctx.normalize_registry.get(scalar.name)
            if fn is not None:
                s = fn(s)
        return s, True
    return value, True


def _apply_scalar_parse(
    ctx: WalkContext,
    scalar: ScalarDef,
    value: Any,
    key: str,
    errors: ValidationErrors,
) -> Any:
    shape = scalar.structured_json_type()
    if shape:
        return _parse_structured_json(ctx, scalar, shape, value, key, errors)
    if not scalar.has_custom_parse:
        return value
    fn = ctx.parse_registry.get(scalar.name)
    if fn is None:
        return value
    parsed, errs = fn(value if isinstance(value, str) else str(value))
    if errs:
        errors[key] = list(errs)
        return value
    if scalar.primitive == "Int":
        coerced, ok = coerce_int(parsed, False)
        return coerced if ok else value
    return parsed


def _parse_structured_json(
    ctx: WalkContext,
    scalar: ScalarDef,
    shape: str,
    value: Any,
    key: str,
    errors: ValidationErrors,
) -> Any:
    """Parse a value of a JSON object or array scalar
    (``ScalarDef.structured_json_type``) into that dict or list, the value
    every generated type holds.

    A string is the value's JSON text. When the scalar has a registered parser
    (Generic.StringMap), the text, or the JSON text of a dict or list, goes
    through it and its canonical text is decoded; a parser error is the result.
    Text that does not decode to the declared shape is kept as it is, for
    validation to report.
    """
    text = value if isinstance(value, str) else None
    fn = ctx.parse_registry.get(scalar.name) if scalar.has_custom_parse else None
    if fn is not None:
        if text is None:
            try:
                text = json.dumps(value, separators=(",", ":"), allow_nan=False)
            except (TypeError, ValueError):
                return value
        parsed, errs = fn(text)
        if errs:
            errors[key] = list(errs)
            return value
        text = parsed
    if text is None:
        return value
    try:
        decoded = json.loads(text)
    except ValueError:
        return value
    return decoded if json_shape_of(decoded) == shape else value


def _apply_builtin(name: str, value: Any, strict: bool) -> tuple[Any, bool]:
    if name == "Int":
        return coerce_int(value, strict)
    if name == "Float":
        return coerce_float(value, strict)
    if name == "Boolean":
        return coerce_bool(value, strict)
    if name in ("String", "ID"):
        return (value, True) if isinstance(value, str) else (value, False)
    return value, True


def _walk_single_field(
    ctx: WalkContext,
    field: FieldDef,
    key: str,
    kind: str,
    value: Any,
    result: dict[str, Any],
    errors: ValidationErrors,
) -> None:
    if kind == "scalar":
        scalar = ctx.schema.scalars.get(field.type_ref.name)
        if scalar is None:
            result[key] = value
            return
        try:
            out, ok = _apply_scalar(ctx, scalar, value)
        except ValueError as exc:
            append_field_error(
                errors,
                key,
                ValidationError(validator="normalize", message=f"invalid {scalar.name}: {exc}"),
            )
            result[key] = value
            return
        if not ok:
            append_field_error(
                errors,
                key,
                ValidationError(validator="type", message=_type_mismatch_message(scalar)),
            )
            result[key] = value
            return
        result[key] = _apply_scalar_parse(ctx, scalar, out, key, errors)
        return
    if kind == "enum":
        result[key] = value
        return
    if kind == "builtin":
        out, ok = _apply_builtin(field.type_ref.name, value, ctx.strict)
        if not ok:
            append_field_error(
                errors,
                key,
                ValidationError(validator="type", message=f"expected {field.type_ref.name} value"),
            )
            result[key] = value
            return
        result[key] = out
        return
    if kind == "type":
        td = ctx.schema.types.get(field.type_ref.name)
        if td is None or not _is_object(value):
            append_field_error(
                errors, key, ValidationError(validator="type", message="expected object value")
            )
            result[key] = value
            return
        nested = walk_type_def(ctx, td, value)
        if _has_errors(nested[1]):
            errors[key] = nested[1]
        result[key] = nested[0]
        return
    if kind == "input":
        td = ctx.schema.inputs.get(field.type_ref.name)
        if td is None or not _is_object(value):
            append_field_error(
                errors, key, ValidationError(validator="type", message="expected object value")
            )
            result[key] = value
            return
        nested = walk_type_def(ctx, td, value)
        if _has_errors(nested[1]):
            errors[key] = nested[1]
        result[key] = nested[0]


def _walk_array_elem(
    ctx: WalkContext,
    field: FieldDef,
    kind: str,
    elem: Any,
    elem_key: str,
    errors: ValidationErrors,
) -> Any:
    if elem is None:
        return None
    if kind == "scalar":
        scalar = ctx.schema.scalars.get(field.type_ref.name)
        if scalar is None:
            return elem
        v, ok = _apply_scalar(ctx, scalar, elem)
        if not ok:
            append_field_error(
                errors,
                elem_key,
                ValidationError(validator="type", message=_type_mismatch_message(scalar)),
            )
            return elem
        return _apply_scalar_parse(ctx, scalar, v, elem_key, errors)
    if kind == "enum":
        return elem
    if kind == "builtin":
        v, ok = _apply_builtin(field.type_ref.name, elem, ctx.strict)
        if not ok:
            append_field_error(
                errors,
                elem_key,
                ValidationError(
                    validator="type", message=f"expected {field.type_ref.name} value"
                ),
            )
            return elem
        return v
    if kind in ("type", "input"):
        defs = ctx.schema.types if kind == "type" else ctx.schema.inputs
        td = defs.get(field.type_ref.name)
        if td is None or not _is_object(elem):
            append_field_error(
                errors,
                elem_key,
                ValidationError(validator="type", message="expected object value"),
            )
            return elem
        nested = walk_type_def(ctx, td, elem)
        if _has_errors(nested[1]):
            errors[elem_key] = nested[1]
        return nested[0]
    return elem


def _walk_array_field(
    ctx: WalkContext,
    field: FieldDef,
    key: str,
    kind: str,
    value: Any,
    result: dict[str, Any],
    errors: ValidationErrors,
) -> None:
    if not isinstance(value, list):
        append_field_error(
            errors, key, ValidationError(validator="type", message="expected array value")
        )
        result[key] = value
        return
    if not field.type_ref.is_array_of_arrays:
        result[key] = [
            _walk_array_elem(ctx, field, kind, elem, f"{key}[{index}]", errors)
            for index, elem in enumerate(value)
        ]
        return
    # A list of lists: walk every inner list. A null inner list passes
    # through; validation rejects it.
    out: list[Any] = []
    for index, row in enumerate(value):
        row_key = f"{key}[{index}]"
        if row is None:
            out.append(None)
            continue
        if not isinstance(row, list):
            append_field_error(
                errors, row_key, ValidationError(validator="type", message="expected an array")
            )
            out.append(row)
            continue
        out.append(
            [
                _walk_array_elem(ctx, field, kind, elem, f"{row_key}[{inner_index}]", errors)
                for inner_index, elem in enumerate(row)
            ]
        )
    result[key] = out


def _walk_field(
    ctx: WalkContext,
    field: FieldDef,
    raw: dict[str, Any],
    result: dict[str, Any],
    errors: ValidationErrors,
) -> None:
    key = field_key(field)
    present = key in raw
    value = raw.get(key) if present else None

    if not present:
        if field.default_value is None:
            return
        if field.type_ref.is_array:
            return
        kind = _resolve_ref_kind(ctx.schema, field.type_ref)
        if kind in ("type", "input"):
            return
        scalar = ctx.schema.scalars.get(field.type_ref.name) if kind == "scalar" else None
        v, ok = apply_default(field, scalar)
        if not ok:
            append_field_error(
                errors,
                key,
                ValidationError(
                    validator="default",
                    message="default value is not parseable for declared type",
                ),
            )
            return
        result[key] = v
        return

    if value is None:
        result[key] = None
        return

    kind = _resolve_ref_kind(ctx.schema, field.type_ref)
    if field.type_ref.is_array:
        _walk_array_field(ctx, field, key, kind, value, result, errors)
        return
    _walk_single_field(ctx, field, key, kind, value, result, errors)


def walk_type_def(
    ctx: WalkContext, td: TypeDef, raw: dict[str, Any]
) -> tuple[dict[str, Any], ValidationErrors]:
    result: dict[str, Any] = {}
    errors = new_validation_errors()
    known: set[str] = set()
    for field in td.fields:
        key = field_key(field)
        known.add(key)
        _walk_field(ctx, field, raw, result, errors)
    if ctx.strict:
        for k in raw.keys():
            if k in known:
                continue
            append_field_error(
                errors, k, ValidationError(validator="unknown_field", message="unknown field")
            )
    return result, errors


__all__ = ["WalkContext", "walk_type_def", "field_key"]
